cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesDataset.java

```diff
@@ -0,0 +1,184 @@
+
+  /**
+   * Get the preferences for the entity id
+   *
+   * @param entityId the entity id to get the preferences from
+   * @return the map which contains the preferences
+   */
+  public Map<String, String> getProperties(EntityId entityId) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        return getConfigProperties(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES));
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        return getConfigProperties(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()));
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        return getConfigProperties(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()));
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        return getConfigProperties(programId.getNamespace(),
+                                   getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                                   programId.getType().getCategoryName(), programId.getProgram()));
+      default:
+        return new HashMap<>();
+    }
+  }
+
+  public Map<String, String> getResolvedProperties(EntityId entityId) {
+    // if it is instance level get the properties and return
+    if (entityId.getEntityType().equals(EntityType.INSTANCE)) {
+      return getProperties(entityId);
+    }
+
+    // get properties from current level
+    Map<String, String> properties = getProperties(entityId);
+    // if the entity id has a parent id, get the preference from its parent
+    if (entityId instanceof ParentedId) {
+      properties.putAll(getProperties(((ParentedId) entityId).getParent()));
+    } else {
+      // otherwise it is a namespace id, which we want to look at the instance level
+      properties.putAll(getProperties(new InstanceId("")));
+    }
+    return properties;
+  }
+
+  public void setProperties(EntityId entityId, Map<String, String> propMap) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        setConfig(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES), propMap);
+        break;
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        setConfig(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()), propMap);
+        break;
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        setConfig(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()), propMap);
+        break;
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        setConfig(programId.getNamespace(),
+                  getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                  programId.getType().getCategoryName(), programId.getProgram()), propMap);
+        break;
+      default:
+        // do nothing
```

Comment:
should you log a warning or throw an exception if someone tries to set properties on a different type of entity?
Otherwise, if the method returns, they will be left believing that the operation was successful.

Reply:
Exception is now thrown for unsupported entity type